cpp/doc/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,747 @@
+     Output storage is allocated outside the API and an output iterator is passed as an argument. 
+
+The main disadvantages of this type of API are
+
+  1. Header-only APIs can increase compilation time for code that depends on them.
+  2. Some users (especially the cuSpatial Python API) may prefer a cuDF-based API.
+
+The column-based C++ API is a simple layer above the header-only API. This approach protects
+column-based API users from the disadvantages while maintaining the advantages for users of the
+header-only API.
+
+### Input/Output Style
+
+All array inputs and outputs are iterator type templates to enable generic application of the APIs.
+An example function is helpful.
+
+```C++
+template <class LonLatItA,
+          class LonLatItB,
+          class OutputIt,
+          class Location = typename std::iterator_traits<LonLatItA>::value_type,
+          class T        = typename Location::value_type>
+OutputIt haversine_distance(LonLatItA a_lonlat_first,
+                            LonLatItA a_lonlat_last,
+                            LonLatItB b_lonlat_first,
+                            OutputIt distance_first,
+                            T const radius               = EARTH_RADIUS_KM,
+                            rmm::cuda_stream_view stream = rmm::cuda_stream_default);
+```
+
+There are a few key points to notice.
+
+  1. The API is very similar to STL algorithms such as `std::transform`.
+  2. All array inputs and outputs are iterator type templates. 
+  3. Longitude/Latitude data is passed as array of structures, using the `cuspatial::lonlat_2d`
+     type (include/cuspatial/vec_2d.hpp). This is enforced using a `static_assert` in the function
+     body.
+  4. The `Location` type is a template that is by default equal to the `value_type` of the input
+     iterators.
+  5. The floating point type is a template (`T`) that is by default equal to the `value_type` of
+     `Location`.
+  6. The iterator types for the two input ranges (A and B) are distinct templates. This is crucial
+     to enable composition of fancy iterators that may be different types for A and B.
+  7. The size of the input and output ranges in the example API are equal, so the start and end of
+     only the A range is provided (`a_lonlat_first` and `a_lonlat_last`). This mirrors STL APIs.
+  8. This API returns an iterator to the element past the last element written to the output. This
+     is inspired by `std::transform`, even though as with `transform`, many uses of 
+     `haversine_distance` will not need this returned iterator.
+  9. All APIs that run CUDA device code (including Thrust algorithms) or allocate memory take a CUDA
+     stream on which to execute the device code and allocate memory.
+
+## Iterator requirements
+
+All input and output iterators must be device-accessible with random access. They must satisfy the
+requirements of C++ [LegacyRandomAccessIterator][LinkLRAI]. Output iterators must be mutable.
+
+## Multiple Return Values
```

Comment:
It is useful for both APIs. Perhaps I should reiterate that the header-only API should return via output iterator where possible. However when the API needs to allocate the results, it must use return(s).

Reply:
I think it is confusing for me because you have grouped the `libcudf` and `header only` ## Multiple Return Values together, and you also seem to switch between the two frameworks while you talk about handling multiple return values. As a developer who is either interested in `libcudf` or interested in contributing to `header only`, I'd like a little more predictability about which paragraph is for me. This is a small detail though, mostly the document reads well, and I'm not asking you to change anything more.

Reply:
Thanks for the feedback. Let mesee what I can do.

Reply:
> As a developer who is either interested in libcudf or interested in contributing to header only, I'd like a little more predictability about which paragraph is for me.

For what it's worth, I expect anyone contributing to libcuspatial will likely be contributing to both.

Reply:
OK, I found that I could clean this up by using parallel structure between the column-based and header-only API sections. I hope this makes it clearer for the reader.